### DISCUSSION TITLE
Using zigbee --> routers <-- with zigbee2mqtt (i.e not coordinator)

### DISCUSSION THREAD

**Discussion by Gronis** [Koenkk/zigbee2mqtt#7006] [General]
I was planning to use 3x CC2531, one as a coordinator and two as routers. My idea was to use both routers as tcp/ip to zigbee bridges essensially, and by using 2 of them, I would get redundancy if one of them dies/goes offline. The coordinator would basically only be there to manage the network.

I now realise that I cannot use zigbee2mqtt with a router firmware. Why is this not possible? It would be perfect to have 1 coordinator to manage the network, and then have 2 routers to be able to send and receive messages on the zigbee network from/to tcp/ip. I'm starting to think this is not possible with the router firmware.

I get the following error when trying to start zigbee2mqtt with a router based firmware 
```log
Error: SREQ '--> SYS - osalNvWrite - {"id":143,"len":1,"offset":0,"value":{"type":"Buffer","data":[1]}}' failed with status '(0x02: INVALID_PARAM)' (expected '(0x00: SUCCESS)')
```

How could something like this be achived?

Is the answer one of these?:
1. Is this not possible because of some limit with how zigbee works (i.e routers cannot subscribe to any message, or something similar)
2. Is this not possible because the router firmware implementation simply does not support this kind of feature (but it could if implemented, and it would not be a constraint with the zigbee protocol)?
3. This is already working and I'm just not doing it the right way.

Any ideas on this? Am I the only person that thinks a single bridge between zigbee and tcp/ip stack is to fragile?

This question relate somewhat to #3017, but the idea is to use multiple routers rather than multiple coordinators (which is not possible).

**Comment by Gronis** [comment]
Ok, I kind of got this to work.

If I use 2 CC2531 as coordinators and 1 as router, that router can relay messages to and from both coordinators from/to the rest of the network. It is kind of a hack. You can see the network topology from both coordinators using the same config file and database file.

I still have a single point of failure, the CC2531 router. I'll add more nodes and see if the CC2531 router really is necessary when more stuff is connected. My hope is that some of the "light routers" will connect to one coordinator and the rest of the "light routers" to the other coordinator, so no matter what coordinator is down, any message should go through the network because we have many "light routers" that can relay messages to/from the coordinators.

I still think that using routers as zigbee2mqtt bridge should be possible, but I'll continue to dive into this rabbit hole for now.

<img width="701" alt="Screenshot 2021-04-09 at 18 31 07" src="https://user-images.githubusercontent.com/5390421/114212260-e94dce80-9961-11eb-988d-794e321a8c28.png">
<img width="483" alt="Screenshot 2021-04-09 at 18 31 14" src="https://user-images.githubusercontent.com/5390421/114212263-ea7efb80-9961-11eb-9d51-41728ae919be.png">

**Comment by Gronis** [comment]
Update:

I've added a few lights (4 to be exact) and it works as expected. It seams that a router prefer to connect to the coordinator that it originally paired with (depends on the router, some routers connect to both coordinators, some don't).

If I make sure to pair routers with the coordinator that it is supposed to be mounted closest too, I can use any of the coordinator to relay messages (i.e, if one dies, everything works as normal). 

The setup I tested with works like this:

I have two raspberry pi with identical setup one is coordinator C1, the other is coordinator C2

1. use coordinator C1 and pair a few routers. For each light that I want to control, create a new group and add that light to that group.
2. turn off C1 and copy the data folder (configuration.yaml and database.db)
3. use coordinator C2 and pair a few routers. For each light that I want to control, create a new group and add that light to that group.
4. turn off C2 and copy the data folder back to C1 data folder. 

Now, with the same config, I can use both coordinators at the same time for all the lamps/routers/endpoints (no need to have a backup that goes in if the primary dongle stops working). If I turn off one of the coordinators (unpugging power/usb or turing off zigbee2mqtt of one of the devices) I can still control the lights if I send commands to their individual groups (one light in each group).

It is not that easy to setup, but once it is done, the setup seam very resilient. You don't need to use different base mqtt topics, both instances can use the same one (with the same mqtt server).

The different map setups looks like this (both operating at the same time):

<img width="753" alt="Screenshot 2021-04-09 at 22 37 04" src="https://user-images.githubusercontent.com/5390421/114238761-70606e00-9985-11eb-94ba-fabb1e007540.png">
<img width="879" alt="Screenshot 2021-04-09 at 22 36 11" src="https://user-images.githubusercontent.com/5390421/114238772-72c2c800-9985-11eb-9754-b3a878ecb75f.png">

**Comment by jvanderzande** [comment]
Hi @Gronis ,
I've been searching for a while on this very topic too and hit on your discussion here. Interesting and surprising there are no other comments as it seems there are plenty folk trying to have a redundant setup with multiple RPI's. 
I am about to start trying what you have explained here on by backup system with a similar setup, so do you have any more information to share after using it for a while now?
Cheers.

**Reply by Gronis** [reply]
Hi @jvanderzande, nice to have another person to talk to about this. However, I must disappoint you that my original setup was very much a "lab" setup on a single laptop; So I haven't really been using it as a daily driver. The devices that I used when testing are benched devices. I wasn't really at the point to commit my daily setup to this. 

The next step would be to setup 2 independent raspberry pi's which host their own mqtt cluster. The main problem with this setup is to add or remove devices. I had to turn off one of the Zigbee coordinators to get that working, then you have to copy the configuration from the "online" coordinator to the one you took offline while pairing devices. My idea was to write a small python application which monitors the configuration and do this automatically, but I haven't looked at it any further.

I think I have described to process I used here as detailed as I can (in my previous post). I remember that it is important to make light groups and control those rather than the devices directly. Otherwise the message is sometimes not propagated correctly.

If you want to try to replicate my testing, I would be very interested to see how it is going for you and what kind of problems you run into (if any). But I think to get something like this working as a daily driver, it would be necessary to write some software to overcome some of the quirks like making manual light groups for single lights, copy the configuration between the two coordinator host nodes, and disable one of the coordinators automatically when entering pairing modes etc.

**Comment by Gronis** [comment]
I've had an error about 2 times when the usb stick hangs for my daily driver at this point. The 2nd time happened just now. It is easily fixable by unplugging and plugging back the usb stick. However, this highlights my problem and the availability outage of my whole smarthome like this is, in my opinion, a serious oversight in robustness design of the network.

I'm using deconz and this was the error.
```log
2021-06-16 22:49:11 DEBUG (MainThread) [pydeconz.gateway] HTTP request response: [{'error': {'address': '/lights/7/state/on',
            'description': 'Internal error, 951',
            'type': 901}}]
2021-06-16 22:49:11 DEBUG (MainThread) [pydeconz.api] BridgeBusy, schedule retry /lights/7/state {'on': False}
2021-06-16 22:49:19 DEBUG (MainThread) [pydeconz.gateway] Sending "put" "{'on': False}" to "192.168.1.102 /lights/7/state"
2021-06-16 22:49:19 DEBUG (MainThread) [pydeconz.gateway] HTTP request response: [{'error': {'address': '/lights/7/state/on',
            'description': 'Internal error, 951',
            'type': 901}}]
```

Errors like this are expected and can happen. But when it happens, I want a second coordinator to handle everything while I'm being notified that a gateway has failed. Maybe even an automatics solution could be used where the "faulty" gateway is forced into a reboot which mostly should fix the problem. If errors still happens after a reboot, only then the user is notified.

**Reply by alan-rpi** [reply]
Hi Gronis. I hope you continue to experiment and report back. I'm trying to achieve something similar.

I want a pi without internet connection to communicate using MQTT and Zigbee. I'm currently stuck trying to get the pi working with a CC2531 flashed with router software (pi cannot see the port ttyACM0). I've now seen this post of yours and your comment "I now realise that I cannot use zigbee2mqtt with a router firmware. " - why do you think that is the case?

I did initially (in error) burn the CC2531 with Coordinator firmware but it did not find any existing devices. This was running with a local copy of Mosquitto and Zigbee2MQTT as the production pi will have no internet connection (but the development system does). In production it only needs to talk to 1 zigbee device so I'm not worried about having to sync the 2 systems. 

However, I would also like to send stats via mqtt to another broker and I think that's not going to be possible (I read it is possible to bridge mqtt brokers but that's via Ethernet not via zigbee - I currently do not use wifi at all, the whole house is wired). 

It would be nice to be able to use python on the pi to talk direct to the CC2531 acting as a client (or router) so that it is seen on the main zigbee network as a device. I'm now looking at possible libraries.

Alan

**Reply by Gronis** [reply]
The problem with using CC2531 as router is that you cannot subscribe to messages the same way you can subscribe to messages from a coordinator.

Wired LAN or WIFI LAN works the same way from the perspective of the mqtt broker. Bridging should be entirely possible but I have no experience in this.

**Reply by alan-rpi** [reply]
> The problem with using CC2531 as router is that you cannot subscribe to messages the same way you can subscribe to messages from a coordinator.

Thanks for pointing that out. I could live with that (i.e. one way communications)

**Reply by alan-rpi** [reply]
Hi Gronis 
> The problem with using CC2531 as router is that you cannot subscribe to messages the same way you can subscribe to messages from a coordinator.

Actually I now fully understand what you mean and that's a real shame. Koenkk has also just confirmed that a CC2531 only uses the USB for power - i.e. no communication to the Pi.

It's now back to the drawing board unless Koenkk writes a hybrid firmware: either router that can communicate or coordinator where the coordinator function can be switched off. 

One alternative is PVTO's firmware but I would need to use the UART which I understands means I cannot use the Pi headless as the console needs to be disabled.

Another alternative is Xbee but I read there are issues linking it to a Z2M network.

**Reply by Gronis** [reply]
Hi @alan-rpi 

Yes this is the realisation I got to as well.

Regarding UART, you can purchase cheep USB2UART adapters so you can connect UART to the raspberry pi via USB if you already use the UART port for CLI on your pi. How would PVTO's firmware solve this though? If this work, I would say this is the superior solution but I havn't played around with PVTO's router firmware.

What are the issues for Xbee when linking it to Z2M network?

**Reply by alan-rpi** [reply]
Koenkk confirms that the router only uses the USB for power - no communication to the host. Shame, as I can think of many situations were someone would like to use a Pi (e.g. a Zero) as an End Device and communicate easily using zigbee2mqtt. [I have mosquitto + zigbee2mqtt and lots of other HA software running on a Pi4 and that's the coordinator].

I did consider a Xbee or a CC2530 with PVTO's firmware but both use the Pi UART to communicate and that apparently means disabling the console. That would be awkward as I use Ethernet and Putty on a PC for development but in production want to install in a remote location with out Ethernet. Lugging a screen and keyboard to the installed location and using the HDMI port and USB would be awkward as well.

**Reply by Gronis** [reply]
Why do you need to use Pi UART to communicate with a Xbee or CC2530? You can use a usb port for that with a <a href="https://www.amazon.com/Adapter-Serial-Converter-Development-Projects/dp/B075N82CDL/ref=sr_1_4?dchild=1&keywords=usb+to+uart&qid=1626731720&sr=8-4">USB to UART/Serial adapter</a>.

I still don't understand how PVTO's firmware could utilize UART on a CC2530 or Xbee to receive and send zigbee messages. Is <a href="https://ptvo.info/zigbee-configurable-firmware-features/uart/">this</a> what you are refering to?

**Reply by alan-rpi** [reply]
Hi Robin. 
> I still don't understand how PVTO's firmware could utilize UART on a CC2530 or Xbee to receive and send zigbee messages. Is [this](https://ptvo.info/zigbee-configurable-firmware-features/uart/) what you are refering to?

Exactly. As far as I can see, UART is the only option to send my choice of topic and data. All the other options (from memory) are for sending data re onboard sensors or connected sensors. Thanks for suggesting the USB to UART/Serial adapter. 
 
In my project I've hit another communications problem. I'm programming a Pico microprocessor to gather sensor data and I want to send the data back to the Pi when the Pi requests it (i.e. Pico is the Device [Slave in Non PC language]). Micropython on the Pico [aka RP2] does not support Device mode, only Control [Master] mode. Applies to I2C and SPI. May have to switch to an Arduino.
